src/Compiler/Microsoft.AspNetCore.Razor.Language/test/IntegrationTests/ComponentCodeGenerationTestBase.cs

```diff
@@ -1672,6 +1672,115 @@ @typeparam T
+
+public class MyComponent : ComponentBase
+{
+    [Parameter]
+    public string Value { get; set; }
+}"));
+
+        // Act
+        var generated = CompileToCSharp(@"<MyComponent Value=""Hello"" />");
+
+        // Assert
+        AssertDocumentNodeMatchesBaseline(generated.CodeDocument);
+        AssertCSharpDocumentMatchesBaseline(generated.CodeDocument);
+        CompileToAssembly(generated);
+    }
+
+    [IntegrationTestFact]
+    public void AddComponentParameter_WithNameof()
+    {
+        // Arrange
+        AdditionalSyntaxTrees.Add(Parse(@"
+using System;
+using Microsoft.AspNetCore.Components;
+
+namespace Test
+{
+    public class MyComponent : ComponentBase
+    {
+        [Parameter]
+        public string Value { get; set; }
+    }
+}"));
+
+        // Act
+        var generated = CompileToCSharp(@"
+<MyComponent Value=""Hello"" />
+@code {
+    public string nameof(string s) => string.Empty;
+}");
+
+        // Assert
+        AssertDocumentNodeMatchesBaseline(generated.CodeDocument);
+        AssertCSharpDocumentMatchesBaseline(generated.CodeDocument);
+        CompileToAssembly(generated, designTime ? [] : [
+                    // (21,55): error CS0120: An object reference is required for the non-static field, method, or property 'MyComponent.Value'
+                    //             __builder.AddComponentParameter(1, nameof(global::Test.MyComponent.
+                    Diagnostic(ErrorCode.ERR_ObjectRequired, "global::Test.MyComponent.\r\n#nullable restore\r\n#line (1,14)-(1,19) \"x:\\dir\\subdir\\Test\\TestComponent.cshtml\"\r\nValue").WithArguments("Test.MyComponent.Value").WithLocation(21, 55)
+            ]);
+    }
+
+    [IntegrationTestFact]
+    public void AddComponentParameter_EscapedComponentName()
+    {
+        // Arrange
+        AdditionalSyntaxTrees.Add(Parse(@"
+using System;
+using Microsoft.AspNetCore.Components;
+
+namespace Test
+{
+    public class @int : ComponentBase
+    {
+        [Parameter]
+        public string Value { get; set; }
+    }
+}"));
+
+        // Act
+        var generated = CompileToCSharp(@"<int Value=""Hello"" />");
+
+        // Assert
+        AssertDocumentNodeMatchesBaseline(generated.CodeDocument);
+        AssertCSharpDocumentMatchesBaseline(generated.CodeDocument);
+        CompileToAssembly(generated);
+    }
+
+    [IntegrationTestFact]
+    public void AddComponentParameter_DynamicComponentName()
+    {
+        // Arrange
+        AdditionalSyntaxTrees.Add(Parse(@"
+using System;
+using Microsoft.AspNetCore.Components;
+
+namespace Test
+{
+    public class dynamic : ComponentBase
+    {
+        [Parameter]
+        public string Value { get; set; }
+    }
+}"));
+
+        // Act
+        var generated = CompileToCSharp(@"<dynamic Value=""Hello"" />");
+
+        // Assert
+        AssertDocumentNodeMatchesBaseline(generated.CodeDocument);
+        AssertCSharpDocumentMatchesBaseline(generated.CodeDocument);
+        CompileToAssembly(generated);
+    }
```

Comment:
Consider testing a component that is a nested class

Reply:
Hmm, I don't think it's possible to actually reference a component that is in a nested class? I just tried and couldn't make it work.

Reply:
You're right, I forgot but I also tried previously and it seems impossible.